### DISCUSSION TITLE
Custom ESP8266 board fails on boot after flashing

### DISCUSSION THREAD

**Discussion by besi** [micropython/micropython#14279] [ESP8266] [answered]
### Checks

- [X] I agree to follow the MicroPython [Code of Conduct](https://github.com/micropython/micropython/blob/master/CODEOFCONDUCT.md) to ensure a safe and respectful space for everyone.

- [X] I've searched for [existing issues](https://github.com/micropython/micropython/issues) matching this bug, and didn't find any.

### Port, board and/or hardware

esp8266

### MicroPython version

ESP8266_GENERIC-FLASH_1M-20240222-v1.22.2

### Reproduction

I made a [custom pcb](https://github.com/besi/esp8266-pcb) with an ESP8266EX.

I did attempt to flash it using the [latest Micropython 1.22.2](https://micropython.org/resources/firmware/ESP8266_GENERIC-20240222-v1.22.2.bin).

After flashing without any errors I connect to the device using `picocom`:

And I get an infinite error message. 
I am using a powered USB-HUB and a cable that has worked without any issues with dozens of projects.

### Expected behaviour

I can flash the device and boot into a Micropython REPL

### Observed behaviour

Now I do get this output when flashing. So it appears as if everything works out ok.
```
Downloading from https://micropython.org/resources/firmware/ESP8266_GENERIC-20240222-v1.22.2.bin
Writing to /var/folders/v5/d9y1qs_16nv9x5bsvjp1dv480000gn/T/tmpnjcoh5iq/ESP8266_GENERIC-20240222-v1.22.2.bin
100%/Applications/Thonny.app/Contents/Frameworks/Python.framework/Versions/3.10/bin/python3.10 -u -m esptool --port /dev/cu.usbserial-1412440 --chip esp8266 --baud 115200 write_flash --flash_mode dout --flash_size detect --erase-all 0x0 /var/folders/v5/d9y1qs_16nv9x5bsvjp1dv480000gn/T/tmpnjcoh5iq/ESP8266_GENERIC-20240222-v1.22.2.bin
esptool.py v4.6.2
Serial port /dev/cu.usbserial-1412440
Connecting....
Chip is ESP8266EX
Features: WiFi
Crystal is 26MHz
MAC: 4c:eb:d6:fc:80:17
Uploading stub...
Running stub...
Stub running...
Configuring flash size...
Auto-detected Flash size: 16MB
Erasing flash (this may take a while)...
Chip erase completed successfully in 36.5s
Flash params set to 0x0390
Compressed 642080 bytes to 426778...
Writing at 0x00000000... (3 %)
Writing at 0x00005866... (7 %)
...
Writing at 0x0007d48c... (81 %)
Writing at 0x00082c05... (85 %)
Writing at 0x000896f0... (88 %)
Writing at 0x0008fdb4... (92 %)
Writing at 0x00096739... (96 %)
Writing at 0x0009c790... (100 %)
Wrote 642080 bytes (426778 compressed) at 0x00000000 in 41.2 seconds (effective 124.6 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...
Done!
```

The boot message looks like so happily repeating ad infinitum:

```
picocom -b 74880 /dev/cu.usbserial-1412440

load 0x40100000, len 31212, room 16 
tail 12
chksum 0x17
ho 0 tail 12 room 4
load 0x3ffe8000, len 1060, room 12 
tail 8
chksum 0x12
load 0x3ffe8430, len 1124, room 0 
tail 4
chksum 0xc7
csum 0xc7
rf_cal[0] !=0x05,is 0xFF

 ets Jan  8 2013,rst cause:2, boot mode:(3,6)

load 0x40100000, len 31212, room 16 
tail 12
chksum 0x17
ho 0 tail 12 room 4
load 0x3ffe8000, len 1060, room 12 
tail 8
chksum 0x12
load 0x3ffe8430, len 1124, room 0 
tail 4
chksum 0xc7
csum 0xc7
rf_cal[0] !=0x05,is 0xFF

 ets Jan  8 2013,rst cause:2, boot mode:(3,6)

```

### Additional Information

I did try various options and baud rates with the same result:

- flash mode: dio
- flash mode: keep
- flash mode: qio
- baud rate default: 115200
- baud rate: 38400
- Version 1.21.0
- Preview Version: `20240330-v1.23.0-preview.322.g5114f2c1e`
- 1MB Image: ESP8266_GENERIC-FLASH_1M-20240222-v1.22.2.bin

This is the Flash chip that I used: [W25Q128JVSIQ](https://jlcpcb.com/partdetail/WinbondElec-W25Q128JVSIQ/C97521) ([Datasheet](https://wmsc.lcsc.com/wmsc/upload/file/pdf/v2/lcsc/1811142111_Winbond-Elec-W25Q128JVSIQ_C97521.pdf))
![](https://github.com/besi/esp8266-pcb/raw/main/esp8266-pcb.png)

**Comment by robert-hh** [comment, marked as answer]
The ESP8266 ROM startup code has a problem with 16MB flash. It works up to 4 MB. See the respective discussion in the old MicroPyhton forum. e.g. https://forum.micropython.org/viewtopic.php?f=16&t=10675&p=60240&hilit=16MB#p60240 or issue #6359.

**Reply by besi** [reply]
You're a champ! That worked like a charm:

	esptool.py -p /dev/cu.usbserial-1412440 -c esp8266 -b 460800 write_flash -fm dio --flash_size=16MB 0x0000 ~/Downloads/ESP8266_GENERIC-20240222-v1.22.2.bin 0xffc000 ~/esp/esp-idf/components/esptool_py/esptool/test/images/esp8266_sdk/esp_init_data_default.bin


	esptool.py v4.6.2
	Serial port /dev/cu.usbserial-1412440
	Connecting....
	Chip is ESP8266EX
	Features: WiFi
	Crystal is 26MHz
	MAC: 4c:eb:d6:fc:80:0e
	Uploading stub...
	Running stub...
	Stub running...
	Changing baud rate to 460800
	Changed.
	Configuring flash size...
	Flash will be erased from 0x00000000 to 0x0009cfff...
	Flash will be erased from 0x00ffc000 to 0x00ffcfff...
	Flash params set to 0x0290
	Compressed 642080 bytes to 426777...
	Wrote 642080 bytes (426777 compressed) at 0x00000000 in 10.8 seconds (effective 477.6 kbit/s)...
	Hash of data verified.
	Compressed 128 bytes to 75...
	Wrote 128 bytes (75 compressed) at 0x00ffc000 in 0.1 seconds (effective 17.8 kbit/s)...
	Hash of data verified.
	
	Leaving...
	Hard resetting via RTS pin...

	MicroPython v1.22.2 on 2024-02-22; ESP module with ESP8266
Type "help()" for more information.

	picocom -b 74880bps /dev/cu.usbserial-1412440

Now `port_diag` does not seem to know the flash size but that's a cosmetic issue:
```

import port_diag

FlashROM:
Flash ID: 1840ef (Vendor: ef Device: 4018)
Flash bootloader data:
Byte @2: 02
Byte @3: 90 (Flash size: ? Flash freq: 40MHZ)
Firmware checksum:
size: 642064
md5: b5c0246d25a48de7dcd4cd4d29a7830f
True

Networking:
STA ifconfig: ('0.0.0.0', '0.0.0.0', '0.0.0.0', '208.67.222.222')
AP ifconfig: ('192.168.4.1', '255.255.255.0', '192.168.4.1', '208.67.222.222')
Free WiFi driver buffers of type:
0: 8 (1,2 TX)
1: 0 (4 Mngmt TX(len: 0x41-0x100))
2: 8 (5 Mngmt TX (len: 0-0x40))
3: 4 (7)
4: 7 (8 RX)
lwIP PCBs:
Active PCB states:
Listen PCB states:
TIME-WAIT PCB states:
```

**Reply by wsnfly** [reply]
I also encountered the same problem, how to solve it? That website can no longer be opened, error 404

**Reply by robert-hh** [reply]
If you have a 16M flash, then you have to write `esp_init_data_default.bin` in addition to the firmware to the end of the flash. I have it attached below as a ZIP file. For a 16M flash, the command would be:

    esptool.py --port <serial-port-of-ESP8266> write_flash -fm DIO 0 <firmware.bin> 0xffc000 esp_init_data_default.bin

<firmware.bin> is the MicroPython firmware you want to install. With a 4MB flash change `0xffc000` to `0x3fc000`.
Another typical reasons for devices not starting is the wrong flash mode, given with the option -fm. You may have to change DIO to other flash modes, like DOUT. See `esptool.py write_flash --help`

[esp_init_data_default.zip](https://github.com/user-attachments/files/22427642/esp_init_data_default.zip)
